### ISSUE TITLE
[ENH]: BoundaryNorm should not need to know colormap length

### ISSUE THREAD

**Issue by jklymak** [matplotlib/matplotlib#21911]
### Problem

Currently `BoundaryNorm` requires `ncolors` as an argument, and hence the user needs to know the colormap size to use it.  

Then we have this comment in the code
```python
    # Mapping to the 0-1 interval could have been done via piece-wise linear
    # interpolation, but using integers seems simpler, and reduces the number
    # of conversions back and forth between int and float.
```

and this "note" in the docstring:

```python
"""
        If there are fewer bins (including extensions) than colors, then the
        color index is chosen by linearly interpolating the ``[0, nbins - 1]``
        range onto the ``[0, ncolors - 1]`` range, effectively skipping some
        colors in the middle of the colormap.
"""
```

If we are going to interpolate anyway, we should just map from 0-1 with `len(boundaries)` boundaries like all the other norms.  Then the user doesn't need to specify Ncolors.  In fact, if the user mispecifies the length of the colormap, then the mapping will mysteriously be incorrect.  (Yes there is a fast-path of len(boundaries)-1 == ncolors, but I really don't think `np.digitize` is one of our performance chokepoints)

### Proposed solution

The problem with changing this is tracking down the special cases, and possible with downstream libraries that do something with the Norm.  The other problem is somehow deprecating the argument.

**Comment by timhoffm** [comment]
BoundaryNorm is particular and results in many quirks:

- it must know and match the number of colors in the colormap (this issue)
- it maps to integers, whereas other norms map to [0, 1]
- therefore it relies on indexing into the colormap's lookup table
- to support that in the standard colorizer pipeline, `Colormap.__call__` special-cases int arguments, which leads to odd behavior https://github.com/matplotlib/matplotlib/issues/28198
- I've always struggled with the name `BoundaryNorm`. In my association that draws attentions to the bounds (=outer values). I think a name like `IntervalNorm`, `SegmentedNorm`, `StepNorm`, `StairNorm` or `PiecewiseConstantNorm` would have been better.

On:
``` 
    # [...] using integers seems simpler, and reduces the number
    # of conversions back and forth between int and float.
```
I believe this is a false assumtion. Using integers may be simpler for a direct mapping, but is much harder to use in a colorizer pipeline because of all the special-casing (see above). The argument on conversion efficiency is premature optimization. If calculating the norm mapping is good enough for all other norms we have, why do we need to jump extra hoops to be more efficient in this case?

Given all this, I propose to create a new norm that regularly maps to [0, 1]. and deprecate `BoundaryNorm`. That is much simpler than trying to fix `BoundaryNorm`.

**Comment by story645** [comment]
> Yes there is a fast-path of len(boundaries)-1 == ncolors

I have a lot of code that did that before I learned the cmap.N trick. Which is it remotely feasible to do some kind of delayed evaluation such that if ncolors==None use cmap.N?

> , I propose to create a new norm that regularly maps to [0, 1]. and deprecate BoundaryNorm.

I think trying to teach users to use the new function will be more work than shimming the new behavior under the old function name given the only difference in signature would be the removal of ncolors. 

Also many of your proposed names are specific functions that have the same structure of continuous measure binned on an ordinal scale where I think we want to avoid implying that this norm is only applicable for specific functions and instead be clear that this norm is intended for any data with this structure. ETA: Though in this direction I guess there's `MonotonicNorm` since that's the underlying constraint on the bounds.

**Comment by timhoffm** [comment]
> is it remotely feasible to do some kind of delayed evaluation such that if ncolors==None use cmap.N?

Not really. The relation is essentially `color = cmap(norm(data))`. So the norm does not know about the colormap. An approximate solution would be to wrap that code inside of the colorizer pipeline (i.e. `Colorizer.to_rgba()`) and do the delayed initialization there. But you can't make an isolated `cmap(norm(data))` work for BoundaryNorm without N.

In the end, you have to switch to mapping to [0, 1] anyway if you want to get rid of all the other quirks. And IMO implementing that in a new norm is simpler than trying to migrate `BoundaryNorm` - after all the out changes from integers to [0, 1], which is a massive breaking change.

> many of your proposed names

Name up to bikeshedding. But I think we'll be able to come up with something reasonable.

> I think trying to teach users to use the new function will be more work than shimming the new behavior under the old function name given the only difference in signature would be the removal of ncolors.

I think the work for teaching and user-side changes is comprable for these two:
- "replace `BoundaryNorm(limits, cmap.N)` by `BoundaryNorm(limits)`"
- "replace `BoundaryNorm(limits, cmap.N)` by `StairNorm(limits)`" (as said name t.b.d.)

**Comment by story645** [comment]
> I think the work for teaching and user-side changes is comprable for these two:\n\n\"replace BoundaryNorm(limits, cmap.N) by BoundaryNorm(limits)\"\n\"replace BoundaryNorm(limits, cmap.N) by StairNorm(limits)\" (as said name t.b.d.)

I think a lot of users don't even remember they're supposed to use cmap.N so cmap.N disappearing will phase them less - especially if we include a grace "optional but unused cmap.N" period.  

I'm also thinking of all the tutorials and training data already out there for BoundaryNorm - any new function is gonna hit the discoverability problem and I don't think it's worth doing that here given we already have the discoverable well documented well publicized name for the thing.  Which also `{function type}Name` is gonna hurt the discoverability for all use cases that don't strictly match {function type}.

**Comment by timhoffm** [comment]
My point is that `BoundaryNorm` is fundamentally flawed by not mapping to [0, 1], with a lot of negative impact. If we want to fix this, we have to substantially change what `BoundaryNorm` is and how it behaves, and you cannot hide this from the user. In particular, there exists no clean migration path. You may be able to gloss over the changes for the standard use case in a colorizer pipeline by jumping some extra hoops, but raw `norm(data)` calls cannot be reasonable be migrated, other than with a hard breaking change.


I still claim that the transition is simpler with a replacement than with an evolution of `BoundaryNorm`.

**Comment by jklymak** [comment]
Replacement seems reasonable to me.

**Comment by story645** [comment]
> calls cannot be reasonable be migrated, other than with a hard breaking change.

Folks will have to hard migrate either way: presumably we'd deprecate `BoundaryNorm` so they'd have to swap in the new name wherever they're currently using `BoundaryNorm`. How `BoundaryNorm` maps to colors ([0,1] vs integer) is an architecture/dev concern, sure, but I'm not sure most users care & the story we're going to be trying to sell them is "change `BoundaryNorm` to `NewNorm` in all your code because we changed the internals but it's still the exact same use cases".

If you keep the name, then you throw an API change warning on norm that it now spits out a different value - I don't see how that's any more onerous than a very long deprecation cycle for `BoundaryNorm`.

**Comment by jklymak** [comment]
I'd just soft-deprecate it by removing mention from all examples, but continue to support the old way.
